src/pudl/harvest/harvest.py

```diff
@@ -0,0 +1,1054 @@
+            used when `errors='report'`.
+            By default, the original error is returned.
+
+    Returns:
+        Aggregated dataframe with `by` columns set as the index and
+        (if `errors`='report') an error report with a :class:`pd.Series` of errors
+        (or the value returned by `errorfunc`) for each column where errors occured.
+
+    Examples:
+        >>> import pandas as pd
+        >>> df = pd.DataFrame({
+        ...     'x': [0, 0, 1, 1],
+        ...     'y': pd.Series([2, 2, 2, 3], dtype='Int64')
+        ... })
+        >>> df.index = [0, 0, 0, 1]
+        >>> base = dict(df=df, by='x', aggfuncs={'y': unique})
+        >>> groupby_aggregate(**base, errors='raise')
+        Traceback (most recent call last):
+          ...
+        ValueError: Could not aggregate y: Not unique.
+        >>> groupby_aggregate(**base, errors='coerce')
+              y
+        x ...
+        0     2
+        1  <NA>
+        >>> result, report = groupby_aggregate(**base, errors='report')
+        >>> result
+              y
+        x ...
+        0     2
+        1  <NA>
+        >>> report['y']
+        x
+        1    Not unique.
+        Name: y, dtype: object
+        >>> errorfunc = lambda x, e: as_dict(x)
+        >>> result, report = groupby_aggregate(**base, errors='report', errorfunc=errorfunc)
+        >>> report['y']
+        x
+        1    {0: [2], 1: [3]}
+        Name: y, dtype: object
+    """
+    data_columns = [col for col in df.columns if col not in by]
+    dtypes = {col: df[col].dtype for col in data_columns}
+    if errors == "report":
+        reports = {}
+        # Prepare columns for error objects
+        df = df.astype({col: object for col in data_columns})
+    if errors == "raise":
+        aggfuncs = {
+            col: try_aggfunc(
+                func, method="raise", error=f"Could not aggregate {col}: {{e}}"
+            )
+            for col, func in aggfuncs.items()
+        }
+    elif errors == "coerce":
+        aggfuncs = {
+            col: try_aggfunc(func, method="coerce") for col, func in aggfuncs.items()
+        }
+    elif errors == "report":
+        aggfuncs = {
+            col: try_aggfunc(func, method="insert", error=errorfunc)
+            for col, func in aggfuncs.items()
+        }
+    if data_columns:
+        result = df.groupby(by).aggregate(aggfuncs)
+    else:
+        # groupby.aggregate drops index when aggfuncs is empty
+        result = df[by].drop_duplicates().set_index(by)
+    for col in data_columns:
+        ds = result[col]
+        if errors == "report":
+            is_error = ds.apply(isinstance, args=(ValueError,))
+            report = ds[is_error]
+            if not report.empty:
+                reports[col] = report
+                ds = ds.where(~is_error)
+        # Enforce original data types, which nulls and errors may have changed
+        if str(dtypes[col]) != str(ds.dtype):
+            ds = ds.astype(dtypes[col])
```

Comment:
For performance. The data types of key columns do not change because they are not modified by nulls or error objects. Data columns only need their type changed back if it changed in the process. `if str(dtypes[col]) != str(ds.dtype)` is to prevent making an unnecessary copy of the column when the column already has the right type.